scripts/apps/knowledge-base/knowledge-base-page.tsx

```diff
@@ -0,0 +1,369 @@
+import React from 'react';
+import {noop, omit} from 'lodash';
+import ReactPaginate from 'react-paginate';
+
+import {ListItem, ListItemColumn, ListItemActionsMenu} from 'core/components/ListItem';
+import {PageContainer, PageContainerItem} from 'core/components/PageLayout';
+import {KnowledgeItemViewEdit} from './knowledge-item-view-edit';
+import {
+    SidePanelHeader,
+    SidePanel,
+    SidePanelHeading,
+    SidePanelTools,
+    SidePanelContent,
+    SidePanelContentBlock
+} from 'core/components/SidePanel';
+import {IFormGroup, IFormField} from './generic-form/interfaces/form';
+import {FormViewEdit} from './generic-form/from-group';
+import {SearchBar} from 'core/ui/components';
+import {Button} from 'core/ui/components/Nav';
+import {SortBar, ISortFields} from 'core/ui/components/SortBar';
+import {Positioner} from 'superdesk-ui-framework';
+import {connectCrudManager, ICrudManager} from 'core/helpers/CrudManager';
+import {TagLabel} from 'core/ui/components/TagLabel';
+import {connectServices} from 'core/helpers/ReactRenderAsync';
+import {IDefaultApiFields} from '../../types/RestApi';
+import {getFormGroupForFiltering} from './generic-form/get-form-group-for-filtering';
+import {getFormFieldsRecursive, getFormFieldPreviewComponent} from './generic-form/form-field';
+
+interface IState {
+    itemInPreview?: string;
+    newItem: null | {[key: string]: any};
+    filtersOpen: boolean;
+    filterValues: {[key: string]: any};
+    searchValue: string;
+    loading: boolean;
+}
+
+interface IProps {
+    formConfig: IFormGroup;
+    renderConceptItemRow(item: IKnowledgeBaseItem): JSX.Element;
+
+    // connected
+    conceptItems?: ICrudManager<IKnowledgeBaseItem>;
+    modal?: any;
+}
+
+export interface IKnowledgeBaseItem extends IDefaultApiFields {
+    name: string;
+    labels?: Array<string>;
+    language: string;
+    definition: string;
+
+    // http://cv.iptc.org/newscodes/cpnature/
+    cpnat_type: 'cpnat:abstract' | 'cpnat:event' | 'cpnat:geoArea'
+        | 'cpnat:object' | 'cpnat:organisation' | 'cpnat:person' | 'cpnat:poi';
+}
+
+class KnowledgeBasePageComponent extends React.Component<IProps, IState> {
+    previewInEditMode: boolean;
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
```

Comment:
not sure if having all logic in component is better than using redux, imo it makes it harder to reuse some pieces of functionality like filtering/pagination, any idea how to do it using this pattern?

Reply:
Filtering/pagination is made up of 2 pieces: logic and UI. The logic is already in `CrudManager` and I don't see how redux would help. You can connect `CrudManager` to React component already.

Regarding UI, it would be good to have a proper reusable list component too. After implementing `KnowledgeBasePageComponent` I realized there is little code specific to the knowledge base and it's almost a generic list component. I should be able to decouple it completely pretty easily.